Simpler check-typo

gasche opened the pull request:
I wanted this to fix [this particular comment](https://github.com/ocaml/ocaml/pull/1905#discussion_r202782208) of @dra27 in the review of #1905. If the present PR is accepted, I will rebase #1905 on top of it.

gasche committed:
check-typo: simplify implementation by deconstructing the built-in exceptions

gasche committed:
check-typo: make (--check-prune .git) work as expected

gasche committed:
rework the relation between .gitattributes and ./tools/check-typo

Before this patch, check-typo is directed by a single git attribute,
ocaml-typo, which is used as a key to set a value:

    ocaml-typo=long-line,missing-header

(the value here is `long-line,missing-header`, and the code splits the
comma later)

This model is very fragile because .gitattributes does not allow to
give attribute keys a collecting/aggregating semantic: each new
setting of the key removes the previous setting, instead of adding to
them. For example,

    testsuite/tests/**                      ocaml-typo=missing-header
    testsuite/tests/win-unicode/*.ml        ocaml-typo=utf8

and

    testsuite/tests/win-unicode/*.ml        ocaml-typo=utf8
    testsuite/tests/**                      ocaml-typo=missing-header

are not equivalent, and instead of using either one we would introduce
redundancy for robustness:

    testsuite/tests/**                      ocaml-typo=missing-header
    testsuite/tests/win-unicode/*.ml        ocaml-typo=missing-header,utf8

With this patch, we switch to a model where each ocaml-typo setting is
its own attribute, of the form `typo.<<attribute>>`. The lines above
would be written, in either order:

    testsuite/tests/**                      typo.missing-header
    testsuite/tests/win-unicode/*.ml        typo.utf8

Not only does this approach make our .gitattributes more robust, it
allows for a more fine-grained treatment of the "unused-prop"
marker. This was used as an attribute to say: don't make it in an
error if the given typo-rule is in fact respected (by default, opting
out of a typo-rule gives an error if the typo-rule is respected). But
because of the single-key nature of ocaml-typo, unused-prop would
range over all settings, not just one of them. For example

    emacs/caml.el ocaml-typo=long-line,unused-prop,missing-header

seems to suggest that 'unused-prop' only qualifies the 'long-line'
rule, but in fact it also ranges over 'missing-header'. In contrast,
with this patch, we write the following:

    emacs/caml.el typo.long-line=may typo.missing-header

the `=may` value setting is used to make an exception to a typo-rule
optional.
Interestingly, most .gitattributes lines worked without extra error
when I turned each unused-prop in a =may setting over the rule just
before, instead of all rules: our checking is now more precise than
before, better capturing the intent of the .gitattributes author.

As I had to rewrite parts of the check-typo code for this, I took the
opportunity to rename a couple variables speaking about SVN (now long
defunct) into more meaningful names:

- `$is_svn` => `$path_in_index`
- `$svnrules` => `$attr_rules`

gasche committed:
check-typo: make sure that 'prune' attributes on non-directories are honored

gasche committed:
check-typo: remove useless gitattributes

gasche commented:
@dra27 where is the use of check-typo (and maybe the pre-commit-hook) documented? If you have committed any documentation in the repository (I could only find a mention in passing in CONTRIBUTING.md), I think that it would be nice to have something in HACKING.adoc -- either the toplevel one, or a new tools/HACKING.adoc.

gasche committed:
Changes entry

dra27 commented:
@gasche - I'm slightly embarrassed to say that while it's on my TODO list, documentation beyond the mention in CONTRIBUTING is all that I've committed so far.

dra27 commented on .gitattributes:
```diff
@@ -16,8 +16,8 @@
-# Don't believe there's a way to wrap lines in .gitattributes
-.gitattributes ocaml-typo=long-line
+# Don't believe there's a way to wrap lines in .gitattributes files
+.gitattributes typo.long-line=may
```
Nit: if we're renaming `ocaml-typo`, perhaps `lint` is better (and still only 4 chars)?

damiendoligez replied:
@dra27 in my eye, the typo checks are much weaker than what a proper lint tool would do.

gasche replied:
(Out of laziness, let's keep the current name for now.)

dra27 commented on .gitattributes:
```diff
@@ -38,68 +38,81 @@
+/.merlin                 typo.missing-header
+/Changes                 typo.utf8 typo.missing-header
+/INSTALL                 typo.missing-header
+/LICENSE                 typo.long-line typo.very-long-line typo.missing-header
```
Surely at this point we could try to subsume long-line and very-long-line into something neater? At the very least, couldn't the Awk script automatically add when `long-line` when `very-long-line` is specified?

damiendoligez replied:
@dra27 I thought it was doing that already? Did I just dream it?

dra27 replied:
No - you definitely need both; it could be done by having the test for `long-line` being between 81 and ~131~132 characters and then `very-long-line` being > 132.

damiendoligez replied:
That sounds like a good idea (modulo the off-by-one error in your description).

dra27 replied:
Oops!

gasche replied:
This is a bit tricky (as in: it requires me to rediscover `awk` again). Given the large latency on my side, I won't touch this feature request and leave it for a future PR.

gasche commented:
@dra27: I'm hoping that you will eventually review this PR (and later #1905). There is nothing urgent about it, so feel free to prioritize other things.

dra27 commented:
I had started to review it! I think the changes are great, just that going through it was taking time - I'll definitely come to it when I'm back from vacation in the second week of Aug.

gasche commented:
Thanks for the update, and have nice vacations :-)

gasche commented:
(gentle ping)

gasche commented:
Would maybe someone else be interested in looking at this? @nojb, @Octachron, @trefis?

dra27 commented on .gitattributes:
```diff
@@ -16,8 +16,8 @@
 
-# Don't believe there's a way to wrap lines in .gitattributes
-.gitattributes ocaml-typo=long-line
+# Don't believe there's a way to wrap lines in .gitattributes files
```
You could make this stronger - there is no way, nor can there ever be, given backwards compatibility

dra27 replied:
(it would be an amusing diversion, for some value of amusing, to lint these files such that *comment blocks* must always be 80 chars or fewer, but that "code" blocks may exceed it)

gasche replied:
I changed it. Note that I think we could now fit within a 80-characters budget, given that the new typo-handling style supports breaking a line by duplicating the path part. But we are still limited by the path length, which reaches 67 for some testsuite files and could conceivably go above 80, so I kept the comment.

dra27 commented on .gitattributes:
```diff
@@ -38,68 +38,81 @@
+*.md                     typo.missing-header
+README*                  typo.missing-header
+*.adoc                   typo.missing-header
+*.adoc                   typo.long-line=may
```
Can the two `*.adoc` lines be merged? It should technically be `*.adoc typo.long-line=may typo.missing-header=may` but I agree with the change where the handling of `missing-header` is concerned.

gasche replied:
I prefer to keep them separate as the first part is part of a "no header for text files" block, while the second is more .adoc-specific.

dra27 commented on .gitattributes:
```diff
@@ -38,68 +38,81 @@
 
-/config/gnu              ocaml-typo=prune
+asmcomp/*/emit.mlp       typo.tab=may typo.long-line=may
+asmcomp/power/NOTES.md   typo.missing-header
```
I can't see a problem with `*.md typo.missing-header` with the `*.adoc` pattern above?

gasche replied:
Indeed, thanks!

dra27 commented on .gitattributes:
```diff
@@ -38,68 +38,81 @@
+testsuite/tests/lib-unix/win-stat/fakeclock.c           typo.missing-header=false
+testsuite/tests/misc-unsafe/almabench.ml                typo.long-line
+testsuite/tests/tool-toplevel/strings.ml                typo.utf8
+testsuite/tests/typing-unboxed-types/test.ml.reference* typo.white-at-eof
```
This can be removed, given the new prune on reference files below

dra27 replied:
```suggestion
```

damiendoligez replied:
@dra27 did you miss the `*` at the end of the pattern?

dra27 replied:
You're right, I had! In which case, it should be `typo.prune`, but perhaps this can be simplified by altering the tests. The following files:

```
tests/typing-unboxed-types/test.ml.reference-flat
tests/typing-unboxed-types/test.ml.reference-noflat
tests/translprim/array_spec.compilers.reference.flat
tests/translprim/module_coercion.compilers.reference.flat
tests/translprim/module_coercion.compilers.reference.no-flat
tests/translprim/array_spec.compilers.reference.no-flat
```
are the only ones which don't end `.reference`. Perhaps they should just be renamed to `.flat.reference` and `.no-flat.reference` so that the normal prune rule comes into play?

dra27 commented on .gitattributes:
```diff
@@ -38,68 +38,81 @@
+tools/magic         typo.missing-header
 
-yacc/*.[ch]   ocaml-typo=long-line,very-long-line,unused-prop
+yacc/*.[ch]         typo.long-line typo.very-long-line=may
```
Shouldn't this be `typo.long-line=may` too?

dra27 commented on .gitattributes:
```diff
@@ -38,68 +38,81 @@
-ocamldoc/Changes.txt     ocaml-typo=missing-header
-ocamldoc/ocamldoc.sty    ocaml-typo=missing-header
-ocamldoc/odoc_index.html ocaml-typo=missing-header
+/manual                  typo.prune
```
Can this have a comment above it explaining that this is a temporary state of affairs (i.e. make it clear to a willing volunteer that it would be preferred to bring the manual into the arms of `check-typo` as well)?

dra27 commented on .gitattributes:
```diff
@@ -38,68 +38,81 @@
-otherlibs/win32unix/readlink.c    ocaml-typo=long-line
-otherlibs/win32unix/stat.c        ocaml-typo=long-line
-otherlibs/win32unix/symlink.c     ocaml-typo=long-line
+ocamldoc/**              typo.long-line=may
```
Similarly - is this another attribute that it would be worth admitting our shame of?

damiendoligez replied:
There are so many long lines in ocamldoc that I doubt anyone will ever fix it (except by releasing a tool that obsoletes ocamldoc, wink, wink).

dra27 commented on .gitattributes:
```diff
@@ -38,68 +38,81 @@
-yacc/*.[ch]   ocaml-typo=long-line,very-long-line,unused-prop
+yacc/*.[ch]         typo.long-line typo.very-long-line=may
 
 # Line-ending specifications, for Windows interoperability
```
I like the separation of lint attributes and line-ending controls. Given the overall changes to this file, can we restructure so that all lint instructions (except, perhaps, the one for `.gitattributes` as a very special case) are at the bottom. So can this point onwards be moved below the `Changes` merge driver comment on line 38?

dra27 commented on .gitattributes:
```diff
@@ -38,68 +38,81 @@
+/tools/ci/inria/remove-sinh-primitive.patch typo.white-at-eol typo.missing-header
 
+# ignore auto-generated .depend files
+.depend                  typo.prune
```
This needs to be `.depend*`, or you need to an entry for `.depend.menhir` (the original pattern in check-typo was `.depend*` which would have been to catch `.depend.nt` files as well, I expect)

dra27 commented on tools/check-typo:
```diff
@@ -35,27 +35,9 @@
-# - Any file whose name matches one of the following patterns is
-#   automatically exempt from all rules
-#     *.reference
-#     *.opt_reference
```
Just FTR, `*.opt_reference` was removed with the ocamltest migration (gone since 8c2d90543fce1e5aca1bd9bc796a51f04be48b4f)

dra27 commented on .gitattributes:
```diff
@@ -38,68 +38,81 @@
+
+# prune testsuite reference files
+testsuite/tests/**/*.reference               typo.prune
+testsuite/tests/**/reference                 typo.prune
```
This is used by *just* the `lib-scanf-2` test - can't we just rename the file and get rid of the special case?

damiendoligez replied:
Seconded. That should slightly simplify the test script itself.

gasche replied:
Done.

dra27 commented on tools/check-typo:
```diff
@@ -35,27 +35,9 @@
-#     *.mllib
-#     *.mltop
-#     *.odocl
-#     *.clib
```
Again, just FTR: these were ocamlbuild special cases

dra27 commented on tools/check-typo:
```diff
@@ -88,13 +70,42 @@
+  env $OCAML_CT_GIT_INDEX git check-attr --all $OCAML_CT_CA_FLAG "$1" \
+  | grep -o " typo\\..*$" | sed "s/ typo\\.//g" \
+  | grep -v ": unset" | grep -v ": false" \
+  | sed "s/: set//g" | sed "s/: true//g" | sed "s/: may/?/g"
```
This pipe can be more simply (or at least compactly): `sed -ne '/: \(set\|true\|may\)/s/.* typo\.//p' -e 's/: \(set\|true\)//' -e 's/:.*/?/'`

gasche replied:
My sed-fu is not high enough to understand your version, so I'll stick with dumb pipes for now.

dra27 commented on tools/check-typo:
```diff
@@ -88,13 +70,42 @@
-                                                 ocaml-typo "$2" 2>/dev/null` in
-      *prune*) echo "INFO: pruned directory $2 (ocaml-typo=prune)" >&2; exit 0;;
-      *) exit 3;;
+    case $2,"$(get_attrs "$2" | grep prune)" in
```
Pedantically, do `grep -xF prune`

gasche replied:
I just checked that [FreeBSD](https://www.freebsd.org/cgi/man.cgi?query=grep&sektion=&n=1) and [OSX](https://ss64.com/osx/grep.html) both support `-x` and its longer form `--line-regexp`, so I will use that, thanks.

gasche replied:
(I'm currently confused as to whether `get_attrs` really returns only the value, or may carry some may/must information also on the line.)

gasche replied:
(The may/must information is encoded as a question mark after the attribute name, so indeed `-x` is correct.)

dra27 commented on tools/check-typo:
```diff
@@ -88,13 +70,42 @@
-      *prune*) echo "INFO: pruned directory $2 (ocaml-typo=prune)" >&2; exit 0;;
-      *) exit 3;;
+    case $2,"$(get_attrs "$2" | grep prune)" in
+      *,*prune*|.git,*)
```
The exact line should be `prune` - this is brittle if we happened to add another attribute containing `prune` as a substring? If the grep flags I suggest (text and wholeline) are added, I think the `*`s can go from `*prune*`

dra27 reviewed (commented):
Sorry for the delay reviewing this. I have reviewed very carefully the changes to `.gitattributes` and reckon that all the pertinent logic which was hard-coded in `check-typo` is now transferred to `.gitattributes`.

The changes to the script itself look good to me - perhaps @damiendoligez could cast a quick eye over it for any concerns about coding style?

dra27 reviewed (commented):
Sorry for the delay reviewing this. I have reviewed very carefully the changes to `.gitattributes` and reckon that all the pertinent logic which was hard-coded in `check-typo` is now transferred to `.gitattributes`.

The changes to the script itself look good to me - perhaps @damiendoligez could cast a quick eye over it for any concerns about coding style?

damiendoligez commented on .gitattributes:
```diff
@@ -38,68 +38,81 @@
-/experimental            ocaml-typo=prune
+emacs/*.el               typo.long-line=may
+emacs/caml.el            typo.long-line=may typo.missing-header
+emacs/COPYING            typo.tab typo.non-printing typo.missing-header
```
This one should be `typo.prune`: we're not ever going to change the GPL text, even if we invent new typo checks or whatever.

damiendoligez reviewed (approved):
Looks good!

gasche committed:
check-typo: silence minor warnings

gasche committed:
reorder the testsuite/tests/lib-scanf-2 test structure

The initial motivation is to have a `foo.reference` file
instead of `reference`.

dra27 committed:
Make typo.very-long-line => typo.long-line=may

typo.long-line can still be explicitly set in .gitattributes (and
typo.long-line typo.very-long-line=may will still issue unused if no
lines are more than 80 columns)

dra27 commented:
I just ran the script over the entire repo and found that the `yacc/*.[ch]` was triggering unused warnings (because it was `typo.long-line` rather than `typo.long-line=may`).

I've also pushed a commit which makes it so that `typo.very-long-line` implies `typo.long-line=may`.

This needs rebasing, so feel free to squash the first commit and potentially drop the latter commit when you do!

gasche committed:
address review comments by David and Damien

gasche commented:
I rebased the PR and I'm in the process of going over the review comments. I hope to finish today to get this PR dealt with.

gasche commented:
@dra27 I'm sorry, I am just seeing that you did pushes on your own. (I force-pushed when I rebased.) I fixed the yacc-stuff as well, but I'm happy to integrate the long-line/very-long-line commit in the PR, could you push it again?

dra27 commented:
No apology needed - it's your branch :slightly_smiling_face: Commit re-pushed...

gasche commented:
I believe that I should now have addressed all review comments (thanks!).

gasche committed:
tools/check-typo: add a a "--get-attrs <path>" mode for debug purposes

dra27 reviewed (commented):
One tweak - though mainly a chance to use GitHub's new suggest edit feature :)

gasche commented:
@dra27: the way I read the diff produced by this indeed-intriguing feature, you are propose to remove the `reference.ml* white-at-eof` line. But if I remove it on my local copy, then I do get an alarm:

```
$ ./tools/check-typo testsuite/tests/typing-unboxed-types/test.ml.reference*
testsuite/tests/typing-unboxed-types/test.ml.reference-flat:213.1: [white-at-eof] empty line(s) at EOF
testsuite/tests/typing-unboxed-types/test.ml.reference-noflat:179.1: [white-at-eof] empty line(s) at EOF
```

What's going on here?

damiendoligez commented on testsuite/tests/lib-scanf-2/tscanf2_master.ml:
```diff
@@ -2,8 +2,8 @@
 modules = "tscanf2_io.ml"
-files = "tscanf2_slave.ml"
-reference = "${test_source_directory}/reference"
+files = "tscanf2_worker.ml"
```
These changes are completely unrelated to check-typo (unless you intend to add a style checker). They should be in a separate PR (and probably touch many more files).

gasche commented:
In the interest of not waiting too much on this issue (its discussion in https://github.com/ocaml/ocaml/pull/2096#issuecomment-430376025 bumped it up in my priority list), I'm thinking of merging this PR today or tomorrow.

damiendoligez reviewed (approved):
Looks OK except for the unrelated changes.

shindere commented:
David Allsopp (2018/10/18 07:03 -0700):
> You're right, I had! In which case, it should be `typo.prune`, but perhaps this can be simplified by altering the tests. The following files:
> 
> ```
> tests/typing-unboxed-types/test.ml.reference-flat
> tests/typing-unboxed-types/test.ml.reference-noflat
> tests/translprim/array_spec.compilers.reference.flat
> tests/translprim/module_coercion.compilers.reference.flat
> tests/translprim/module_coercion.compilers.reference.no-flat
> tests/translprim/array_spec.compilers.reference.no-flat
> ```
> are the only ones which don't end `.reference`. Perhaps they should
> just be renamed to `.flat.reference` and `.no-flat.reference` so that
> the normal prune rule comes into play?

Yes, please!

Sorry for having done things the wrong way initially!

gasche commented:
I'm busy with other things and would like to get done with this PR. Is it possible to leave those further suggestions to future PRs?

shindere commented:
Gabriel Scherer (2018/10/18 07:25 -0700):
> I'm busy with other things and would like to get done with this PR. Is
> it possible to leave those further suggestions to future PRs?

Fine with me :)

damiendoligez commented:
@gasche just let me push a commit to rename the reference files before you merge.

damiendoligez committed:
GPR#1910: rename .reference.* files

damiendoligez commented:
Done.

gasche commented:
Thanks! Merged.
